### PR TITLE
Support exclusion of non-application URIs if custom Sampler is used

### LINKED ISSUE #25525: OpenTelemetry: Support exclusion of non-application URIs if custom Sampler is used
### Description

This refers to the `quarkus-opentelemetry` extension.

We have opted for using a custom `io.opentelemetry.sdk.trace.samplers.Sampler` CDI Bean in order to extend the existing sampling logic.
As stated in the documentation, this means that the `quarkus.opentelemetry.tracer.suppress-non-application-uris` option has no effect, so that the logic for suppressing non-application uris has to be added manually to the custom Sampler.

For applications not using a common `/q` prefix for the non-application uris, assembling the URIs and applying the path resolution shows to be quite cumbersome.

It would be helpful, if the
```
List<String> dropNonApplicationUris,
List<String> dropStaticResources
```
information, given to the [TracerRecorder.setupSampler()](https://github.com/quarkusio/quarkus/blob/2205bb989655970f4da72b752ff2c2be7c5103cc/extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerRecorder.java#L139-L140) method is somehow made available for the creation of a custom Sampler.

### Implementation ideas

Instead of checking for a Sampler bean, the `TracerRecorder.setupSampler()` method could check for a bean of a new `SamplerFactory` class, containing a factory method to which the above mentioned `dropNonApplicationUris` and `dropStaticResources` can be supplied by the `TracerRecorder.setupSampler()` method.

### PR DESCRIPTION
**Pull request by mun711**
- Fixes #25525

### CODE REVIEW

**Comment by radcortez** [comment]
Thank you for the PR!

I'm not totally sure if we want to support adding samplers as CDI Beans. I know that we already have a few other things that we configure with CDI Beans, but I believe that was a mistake.

OTel now provides a well-defined configuration to set a lot of stuff, we just need to support it first. I think our goal should be to move to https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md and use `otel.traces.sampler` to set up the Sampler.

**Comment by mun711** [comment]
> Thank you for the PR!
> 
> I'm not totally sure if we want to support adding samplers as CDI Beans. I know that we already have a few other things that we configure with CDI Beans, but I believe that was a mistake.
> 
> OTel now provides a well-defined configuration to set a lot of stuff, we just need to support it first. I think our goal should be to move to https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md and use `otel.traces.sampler` to set up the Sampler.



> Thank you for the PR!
> 
> I'm not totally sure if we want to support adding samplers as CDI Beans. I know that we already have a few other things that we configure with CDI Beans, but I believe that was a mistake.
> 
> OTel now provides a well-defined configuration to set a lot of stuff, we just need to support it first. I think our goal should be to move to https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md and use `otel.traces.sampler` to set up the Sampler.

@radcortez - Thanks for the reply
I checked out the link for opentelemetry autoconfiguration. It seems that such a migration will require quite a few changes. I don't have a full understanding of the quarkus recorder pattern, but I could see that the current opentelemetrysdk is created at that point, and my guess is that we want to somehow replace with AutoConfiguredOpenTelemetrySdk, as well as add some customizations to support custom sampler enhancements and possibly url dropping as a configuration property both for custom and existing samplers.

If we go ahead with the autoconfigure ticket I guess this PR and the relevant issue may be closed as no longer wanted ?

**Comment by radcortez** [comment]
> If we go ahead with the autoconfigure ticket I guess this PR and the relevant issue may be closed as no longer wanted ?

Not sure yet. 

I do know that we want to go with the OTel way to configure things. Since they already have a way to configure all of their APIs we should go with that instead of providing specific ways to configure stuff (CDI Beans). 

We don't even have support configuration as CDI Beans for everything, so we would either need to add it to make it consistent or drop it altogether. This means that we need to somehow migrate / deprecate the current CDI support.

Maybe we just need to offer our custom Sampler that users can extend that excludes the non-application uris out of the box.

File: extensions/opentelemetry/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/TracerUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.trace.data.LinkData;
+import io.opentelemetry.sdk.trace.samplers.Sampler;
+import io.opentelemetry.sdk.trace.samplers.SamplingResult;
+import io.quarkus.opentelemetry.runtime.tracing.DropNamesSampler;
+import io.quarkus.opentelemetry.runtime.tracing.TracerRuntimeConfig;
+import io.quarkus.opentelemetry.runtime.tracing.TracerUtil;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class TracerUtilTest {
```
Comment (radcortez) [inline extensions/opentelemetry/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/TracerUtilTest.java:24]: Please look into the `NonAppEndpoints*Test's` and implement a similar test for a custom sampler in `OpenTelemetrySamplerBeanTest`.

**Review by radcortez — CHANGES_REQUESTED** [review]

File: extensions/opentelemetry/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/TracerUtilTest.java (hunk 1 above)
Comment (mun711) [reply extensions/opentelemetry/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/TracerUtilTest.java]: updated OpenTelemetrySamplerBeanTest to use custom sampler

Comment (radcortez) [reply extensions/opentelemetry/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/TracerUtilTest.java]: The idea was to implement a new test, not change the current one. With this approach, we are not testing the previous scenario when there is no custom CDI Sampler.

Can you please keep the current test and make a new one on the side with the custom CDI Sampler? I'm sorry if I was not clear. Thanks!

Comment (mun711) [reply extensions/opentelemetry/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/TracerUtilTest.java]: sorry, I had the impression that the initial test was testing that a sampler bean is accesible via openTelemetry when defined in the Application context, thus could be replaced with a custom one 
Reverted these changes and added as a separate test

**Review by radcortez — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks!

**Pull request merged**